File: utils/EdgeBlocker.js
/**
 * Win11Debloat - Edge Neutralizer Module (Node.js)
 * Location: utils/EdgeBlocker.js
 *
 * Reads config/edge_block.json and applies:
 *  - IFEO execution redirect on msedge.exe
 *  - Anti-reinstall registry policies
 *  - Active Edge process termination
 */

const fs = require('fs');
const path = require('path');
const engine = require('./DebloatEngine.js');

function applyEdgeBlock() {
    console.log('[*] Reading Edge Block Configuration...');
    let config;
    try {
        config = engine.loadConfig('edge_block.json');
    } catch (err) {
        console.error('[-] ' + err.message);
        process.exit(1);
    }

    // 1. Apply anti-reinstall registry policies
    if (config.registryKeys && Array.isArray(config.registryKeys)) {
        console.log('[*] Applying Anti-Reinstall Registry Policies...');
        config.registryKeys.forEach((item) => {
            const result = engine.setRegistryValue(item.path, item.name, item.type, item.value);
            if (result.success) {
                console.log(`[+] Set ${item.path}\\${item.name} = ${item.value}`);
            } else {
                console.warn(`[!] Failed to set ${item.path}\\${item.name}: ${result.output}`);
            }
        });
    }

    // 2. Apply IFEO redirect (blocks msedge.exe execution permanently)
    if (config.ifeoBlock) {
        console.log('[*] Registering Image File Execution Options (IFEO) Redirect...');
        const ifeoPath = `HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\Image File Execution Options\\${config.ifeoBlock.targetExecutable}`;
        const result = engine.setRegistryValue(ifeoPath, 'Debugger', 'REG_SZ', config.ifeoBlock.debuggerRedirect);
        if (result.success) {
            console.log(`[+] IFEO block placed on ${config.ifeoBlock.targetExecutable}`);
        } else {
            console.warn(`[!] Failed to set IFEO redirect: ${result.output}`);
        }
    }

    // 3. Kill active Edge processes
    console.log('[*] Terminating active Edge processes...');
    const killResult = engine.killProcess('msedge.exe');
    if (killResult.success) {
        console.log('[+] Active Edge processes terminated.');
    } else {
        console.warn(`[!] Edge process termination note: ${killResult.output}`);
    }

    console.log('[+] Edge blocking rules applied successfully.');
}

applyEdgeBlock();

File: utils/EdgeBlocker.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Security.AccessControl;
using System.Security.Principal;
using Microsoft.Win32;

namespace Win11Debloat.Utils
{
    public class EdgeBlocker
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("[*] Starting Deep Edge Removal and Store Lockout Engine...");

            KillEdgeProcesses();
            RunEdgeSetupUninstall();
            ApplyRegistryPolicies();
            ApplyIFEOTrap();
            LockoutEdgeFilesystem();

            Console.WriteLine("[+] Microsoft Edge has been permanently neutralized.");
        }

        private static void KillEdgeProcesses()
        {
            string[] targets = { "msedge", "MicrosoftEdgeUpdate", "identity_helper" };
            foreach (var target in targets)
            {
                foreach (var proc in Process.GetProcessesByName(target))
                {
                    try
                    {
                        proc.Kill();
                        Console.WriteLine($"[+] Terminated process: {target} (PID: {proc.Id})");
                    }
                    catch { }
                }
            }
        }

        private static void RunEdgeSetupUninstall()
        {
            string programFiles = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86);
            string edgeAppDir = Path.Combine(programFiles, "Microsoft", "Edge", "Application");

            if (Directory.Exists(edgeAppDir))
            {
                var setupFiles = Directory.GetFiles(edgeAppDir, "setup.exe", SearchOption.AllDirectories);
                if (setupFiles.Length > 0)
                {
                    string setupPath = setupFiles[0];
                    Console.WriteLine($"[*] Found Edge setup binary: {setupPath}");

                    ProcessStartInfo psi = new ProcessStartInfo
                    {
                        FileName = setupPath,
                        Arguments = "--uninstall --system-level --verbose-logging --force-uninstall",
                        UseShellExecute = false,
                        CreateNoWindow = true
                    };
                    
                    Process p = Process.Start(psi);
                    p?.WaitForExit();
                    Console.WriteLine("[+] Executed force-uninstall routine.");
                }
            }
        }

        private static void ApplyRegistryPolicies()
        {
            Console.WriteLine("[*] Writing anti-reinstallation policies to Registry...");

            try
            {
                using (RegistryKey key = Registry.LocalMachine.CreateSubKey(@"SOFTWARE\Policies\Microsoft\EdgeUpdate", true))
                {
                    key.SetValue("DoNotUpdateToEdgeWithChromium", 1, RegistryValueKind.DWord);
                    key.SetValue("InstallDefault", 0, RegistryValueKind.DWord);
                    key.SetValue("UpdateDefault", 0, RegistryValueKind.DWord);
                    key.SetValue("PreventAppDefaultsAutoOptIn", 1, RegistryValueKind.DWord);
                }

                using (RegistryKey key = Registry.LocalMachine.CreateSubKey(@"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\Microsoft Edge", true))
                {
                    key.SetValue("NoRemove", 0, RegistryValueKind.DWord);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[-] Error applying registry keys: {ex.Message}");
            }
        }

        private static void ApplyIFEOTrap()
        {
            Console.WriteLine("[*] Registering IFEO Debugger Execution Trap...");
            try
            {
                string ifeoPath = @"SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\msedge.exe";
                using (RegistryKey key = Registry.LocalMachine.CreateSubKey(ifeoPath, true))
                {
                    key.SetValue("Debugger", "cmd.exe /c exit", RegistryValueKind.String);
                }
                Console.WriteLine("[+] IFEO trap successfully engaged for msedge.exe.");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[-] Error applying IFEO trap: {ex.Message}");
            }
        }

        private static void LockoutEdgeFilesystem()
        {
            string programFiles = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86);
            string edgeDir = Path.Combine(programFiles, "Microsoft", "Edge");

            if (!Directory.Exists(edgeDir))
            {
                Directory.CreateDirectory(edgeDir);
            }

            try
            {
                Console.WriteLine("[*] Applying Deny ACLs to prevent Microsoft Store / Service writes...");
                DirectoryInfo dInfo = new DirectoryInfo(edgeDir);
                DirectorySecurity dSecurity = dInfo.GetAccessControl();

                SecurityIdentifier everyoneSid = new SecurityIdentifier(WellKnownSidType.WorldSid, null);

                dSecurity.AddAccessRule(new FileSystemAccessRule(
                    everyoneSid,
                    FileSystemRights.FullControl,
                    AccessControlType.Deny
                ));

                dInfo.SetAccessControl(dSecurity);
                Console.WriteLine("[+] Universal DENY permission set on Edge directory.");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[-] Error modifying ACLs: {ex.Message}");
            }
        }
    }
}
